### DISCUSSION TITLE
Button to show midi monitor?

### DISCUSSION THREAD

**Discussion by puffer3** [RomanKubiak/ctrlr#511] [Q&A]
I end up hiding the menu bar when I export panels, but when working on PC's I would find it useful to have a button to bring up the built in midi monitor.  Is there a way to do this?

**Comment by damiensellier** [comment]
There could be one way to call the monitor window is to simulate the "command + m" key combo with a LUA script when a button is pressed but I'm not sure it can do it stock without a library.
I never used key simulation but I know @bijlevel already implemented some keyboard input features in its beautiful Kurzweil sampler panel. 
Please take a look at "simulating a keypress in CTRLR" although I think they used it the other way with keystroke event listeners:
https://github.com/RomanKubiak/ctrlr/discussions/256 

There's the JUCE KeyPress function as well that we use to "enter" "cancel" in popups it's detailed here https://docs.juce.com/master/classKeyPress.html : 

There's something about it in the DEMO - Global keyboard events.panel : 

```
triggerKeyPressEvent = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value)
	delay = panel:getModulator("delayTime"):getValue()
	combo = panel:getLabel("keyPressLabel"):getText()

	sleep (delay * 1000)
	keyPressEvent = KeyPress.createFromDescription(combo)
	native:sendKeyPressEvent(keyPressEvent)
end
```

**Comment by damiensellier** [comment]
I tried with this script on the button value change, but it doesn't work, not sure if this JUCE feature is implemented well in CTRLR:

```
myMethod = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
	
	keyPressEvent = KeyPress.createFromDescription("cmd+M")
	native:sendKeyPressEvent(keyPressEvent)

end
```


the correct c++ would be : 

`KeyPress::KeyPress('m', ModifierKeys::commandModifier, 0)`

I don't know how to convert the syntax for LUA

**Reply by dnaldoog** [reply]
> `KeyPress::KeyPress('m', ModifierKeys::commandModifier, 0)`
> 
> I don't know how to convert the syntax for LUA

Hi Damien,

Hope you are well!

I was mucking around with this too, but got nowhere. This might be how to duplicate that C++ in lua ❓

<pre>
local modKey='m'
KeyPress(string.byte(modKey),ModifierKeys(ModifierKeys.ctrlModifier),0)
</pre>

**Reply by damiensellier** [reply]
Hi John, I'm fine thank you, hope everything is fine on your side.

We started from a "basic" question but it opened doors for me to take a look at luabind and how it's implemented in CTRLR.

To make it simple, I added "m" keypress as a shortcut to the midi monitor
I go with a button assigned to a lua script that replicates the demo panel with simulated keypress : 

```
keyPressEvent = KeyPress.createFromDescription("m")
native:sendKeyPressEvent(keyPressEvent)
```
I don't even get the midi monitor to show up which is strange.
While pressing 'm' key it shows the midi monitor, not by clicking the modulator. 

I thought it was not working because when you click on the button to trigger the keypress "m" it's not just a keypress, but a combo leftclick+M, maybe?

Or maybe it's just that keypress is not fully implemented with luabind, I really don't know. I can already feel my deficiency with not speaking C++ language.

**Reply by Ted-juh** [reply]
I think it's not working the way you want because your command "sendKeyPressEvent" is to pass on keypress events to "outside" Ctrlr. Buttons only have "setCommandToTrigger" attached to them, which is to invoke the button itself with a key press.  

Now I'm looking into "CommandIDs (See CtrlrEditor.h), showMidiMonitor has CommandID 0x2005. Further in CtrlrEditorApplicationCommandHandlers.cpp to get the midimonitor to show up, you do:
owner.getWindowManager().toggle(CtrlrManagerWindowManager::MidiMonWindow, true);

Could there be a way to trigger that with a button? Just stabbing in the dark here :))

**Reply by damiensellier** [reply]
> I think it's not working the way you want because your command "sendKeyPressEvent" is to pass on keypress events to "outside" Ctrlr. Buttons only have "setCommandToTrigger" attached to them, which is to invoke the button itself with a key press.

Oh you're right this is the command from the demo panel and it says to switch app after pressing the button to simulate the keypress on this other app. It makes sense this way.

I'll check today everything about the midiMonWindow in the source code, that could be a door to somthing relevant.

Also the syntax to add buttons in alert box is pretty simple, it could also help understanding the way to make it work :

```
modalWindow:addButton("    OK    ", 1, KeyPress(KeyPress.returnKey),KeyPress())
modalWindow:addButton("Cancel", 0, KeyPress(KeyPress.escapeKey),KeyPress())
```
But in this case, keypress keycodes are pre-registered with luabind...
https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Lua/JuceClasses/LCore.cpp#L362

thanks for the suggestion, it's a great idea @Ted-juh
